contracts/src/registry/EnhancedAccessControl.sol

```diff
@@ -0,0 +1,302 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts (last updated v5.0.0) (access/AccessControl.sol)
+
+pragma solidity ^0.8.20;
+
+import {Context} from "@openzeppelin/contracts/utils/Context.sol";
+import {ERC165} from "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+
+/**
+ * @dev An enhanced version of OpenZeppelin's AccessControl that allows for:
+ * 
+ * - Resource-based roles.
+ * - Root resource override (0x0) - role assignments in the root resource auto-apply to all resources.
+ * - Max 256 roles - stored as a bitmap in uint256.
+ */
+abstract contract EnhancedAccessControl is Context, ERC165 {
+    error EnhancedAccessControlUnauthorizedAccountRole(bytes32 resource, uint256 role, address account);
+    error EnhancedAccessControlUnauthorizedAccountAdminRole(bytes32 resource, uint256 role, address account);
+    error EnhancedAccessControlBadConfirmation();
+    error EnhancedAccessControlLockedRole(bytes32 resource, uint256 role);
+
+    event EnhancedAccessControlRoleAdminChanged(uint256 role, uint256 previousAdminRole, uint256 newAdminRole);
+    event EnhancedAccessControlRoleGranted(bytes32 resource, uint256 role, address account, address sender);
+    event EnhancedAccessControlRolesGranted(bytes32 resource, uint256 roleBitmap, address account, address sender);
+    event EnhancedAccessControlRolesCopied(bytes32 srcResource, bytes32 dstResource, address account, address sender, uint256 roleBitmap);
+    event EnhancedAccessControlRolesRevoked(bytes32 resource, uint256 role, address account, address sender);
+    event EnhancedAccessControlAllRolesRevoked(bytes32 resource, address account, address sender);
+    event EnhancedAccessControlRolesLocked(bytes32 resource, uint256 roleBitmap);
+
+    /** 
+     * @dev admin role that controls a given role. 
+     * RoleId -> AdminRoleId
+     */
+    mapping(uint256 role => uint256 adminRole) private _adminRoles;
```

Comment:
Can we simply say that the upper 128 bits are admin roles and the lower 128 bits are user roles, and thus remove the need to map them explicitly?

Reply:
I think it's possible that a role can be a role as well as an admin role - e.g registrar can be admin roles for setting the renewer admin role, even though registrars are themselves roles.

Reply:
I don't think there's in harm in representing those with separate bits, though? If they're typically used together, the inheriting contract can simply grant or revoke both bits together.